### PR TITLE
Remove all XML help

### PR DESCRIPTION
**Pull request by cormacpayne**
## Description

Fix for issue https://github.com/Azure/azure-powershell/issues/3583

- Remove all XML help from repository
- Add script to build that will generate markdown help, compare it to what exists in the repository, validate the markdown help, and generate XML

---

This checklist is used to make sure that common guidelines for a pull request are followed. You can find a more complete discussion of PowerShell cmdlet best practices [here](https://msdn.microsoft.com/en-us/library/dd878270(v=vs.85).aspx).

- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md).**
- [x] **If changes were made to any cmdlet, the XML help was regenerated using the [platyPSHelp module](https://github.com/Azure/azure-powershell/blob/dev/documentation/platyPSHelp-documentation.md).**
- [x] **If any large changes are made to a service, they are reflected in the respective [change log](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#updating-the-change-log).**

### [General Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#general-guidelines)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).
- [x] The pull request does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/dev/documentation/breaking-changes.md) (unless a major version change occurs in the assembly and module).

### [Testing Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#testing-guidelines)
- [ ] Pull request includes test coverage for the included changes.
- [ ] PowerShell scripts used in tests should do any necessary setup as part of the test or suite setup, and should not use hard-coded values for locations or existing resources.

### [Cmdlet Signature Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#cmdlet-signature-guidelines)
- [ ] New cmdlets that make changes or have side effects should implement `ShouldProcess` and have `SupportShouldProcess=true` specified in the cmdlet attribute. You can find more information on `ShouldProcess` [here](https://gist.github.com/markcowl/338e16fe5c8bbf195aff9f8af0db585d#what-is-the-change).
- [ ] Cmdlet specifies `OutputType` attribute if any output is produced - if the cmdlet produces no output, it should implement a `PassThru` parameter.

### [Cmdlet Parameter Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#cmdlet-parameter-guidelines)
- [ ] Parameter types should not expose types from the management library - complex parameter types should be defined in the module.
- [ ] Complex parameter types are discouraged - a parameter type should be simple types as often as possible. If complex types are used, they should be shallow and easily creatable from a constructor or another cmdlet.
- [ ] Cmdlet parameter sets should be mutually exclusive - each parameter set must have at least one mandatory parameter not in other parameter sets.

### CODE REVIEW

**Comment by cormacpayne** [comment]
@azuresdkci test this please

**Comment by cormacpayne** [comment]
@markcowl resolved issue with wxi file

**Comment by cormacpayne** [comment]
Sign job: http://azuresdkci.cloudapp.net/view/1-AzurePowerShell/job/powershell-sign/936/

**Comment by cormacpayne** [comment]
Sign job: http://azuresdkci.cloudapp.net/view/1-AzurePowerShell/job/powershell-sign/937/

File: documentation/platyPSHelp-documentation.md [hunk 1]
```diff
@@ -1,319 +0,0 @@
-# PlatyPS Help
```
Comment (markcowl) [inline documentation/platyPSHelp-documentation.md]: Let's make sure to add in the new help documentation

File: tools/StaticAnalysis/HelpAnalyzer/HelpAnalyzer.cs [hunk 2]
```diff
@@ -80,54 +80,148 @@ public void Analyze(IEnumerable<string> scopes)
...
+                    }
+                    else
+                    {
+                        ResourceManagerAnalyze(directory, helpLogger, processedHelpFiles, savedDirectory);
+                    }
+                }
+            }
+        }
+
+        private void ServiceManagementAnalyze(
```
Comment (markcowl) [inline tools/StaticAnalysis/HelpAnalyzer/HelpAnalyzer.cs]: ```<nit>AnalyzeMamlHelp</nit>```

File: tools/StaticAnalysis/HelpAnalyzer/HelpAnalyzer.cs [hunk 3]
```diff
@@ -80,54 +80,148 @@ public void Analyze(IEnumerable<string> scopes)
...
-                        .Where (f => IsAssemblyFile(f) && !File.Exists(f + "-Help.xml"));
-                    foreach (var orphanedAssembly in commandAssemblies)
+                    if (directory.Contains("ServiceManagement"))
                     {
-                        helpLogger.LogRecord(new HelpIssue()
+                        ServiceManagementAnalyze(directory, helpLogger, processedHelpFiles, savedDirectory);
+                    }
+                    else
+                    {
+                        ResourceManagerAnalyze(directory, helpLogger, processedHelpFiles, savedDirectory);
```
Comment (markcowl) [inline tools/StaticAnalysis/HelpAnalyzer/HelpAnalyzer.cs]: ```<nit>AnalyzeMarkdownHelp</nit>```

File: tools/GenerateHelp.ps1 [hunk 4]
```diff
@@ -11,21 +11,25 @@ Param(
...
+$HelpFolders = Get-ChildItem "help" -Recurse -Directory | where { $_.FullName -like "*Debug*" -or $_.FullName -like "*Release*" }
 
 # ---------------------------------------------------------------------------------------------
 
 if ($ValidateMarkdownHelp)
 {
     New-Item -Path "$PSScriptRoot\HelpGeneration\Exceptions" -Name NewValidateHelpExceptions.csv -ItemType File -Force | Out-Null
 
-    Add-Content "$PSscriptRoot\HelpGeneration\Exceptions\NewValidateHelpExceptions.csv" "Module,Target,Description"
+    Add-Content "$PSScriptRoot\HelpGeneration\Exceptions\NewValidateHelpExceptions.csv" "Module,Target,Description"
```
Comment (markcowl) [inline tools/GenerateHelp.ps1]: Generally, Exceptions are for expected errors, I think the error csvs are being copied to the build directory, and we should mnake sue of thsi same location  Also, do we have any mechanism for suppressions?

File: build.proj [hunk 5]
```diff
@@ -182,16 +182,17 @@
...
+    <Exec Command="$(PowerShellCommand) -NonInteractive -NoLogo -NoProfile -Command &quot;. $(LibraryToolsFolder)\GenerateHelp.ps1 -ValidateMarkdownHelp -GenerateMamlHelp -BuildConfig $(Configuration) &quot;"/>
     <CallTarget Targets="BuildSetupTest"/>
 	<CallTarget Targets="CodeSignBinaries" Condition=" '$(CodeSign)' == 'true' " />	
 	<CallTarget Targets="BuildSetup"/>
 	<CallTarget Targets="CodeSignInstaller"
                 Condition=" '$(CodeSign)' == 'true' and '$(Scope)' == 'all'" />
-	
 	<Message Importance="high" Text="Running Static Analyser" />
     <CallTarget targets="DependencyAnalysis" />
+    <Exec Command="$(PowerShellCommand) -NonInteractive -NoLogo -NoProfile -Command &quot;. $(LibraryToolsFolder)\CleanupBuild.ps1 -BuildConfig $(Configuration) &quot;"/>
```
Comment (markcowl) [inline build.proj]: Can we make sure the build fails if this fails?  We can do this by settign ContinueOnError="false", similar to this: https://github.com/Azure/azure-powershell/blob/stack-dev/build.proj#L369

File: setup/azurecmdfiles.wxi [hunk 6]
```diff
@@ -462,9 +462,6 @@
                         <Component Id="cmp6BDCB170E1F314882CF58F8F80DE76A6" Guid="*">
                             <File Id="filBFDE1699047559AB8DA89E0F660D35DF" KeyPath="yes" Source="$(var.sourceDir)\ResourceManager\AzureResourceManager\AzureRM.Billing\Microsoft.Azure.Commands.Common.Authentication.dll" />
                         </Component>
-                        <Component Id="cmpED253190CB7DAF0A995567435152DC6A" Guid="*">
```
Comment (markcowl) [inline setup/azurecmdfiles.wxi]: These should not be deleted.  We need the xml help files in the installer, or there will be no online help

File: build.proj (hunk 5 above)
Comment (markcowl) [inline build.proj]: Also, lets make sure that building the installer always happens after generating help

File: tools/GenerateHelp.ps1 [hunk 7]
```diff
@@ -0,0 +1,50 @@
+#Requires -Modules platyPS
+
+[CmdletBinding()]
+Param(
+    [Parameter()]
+    [Switch]$ValidateMarkdownHelp,
+
```
Comment (markcowl) [inline tools/GenerateHelp.ps1]: ```<nit>```remove extra veritcal space.  You should only have a vertical space after a closing brace ```}```,  or the like ```</nit>```

File: tools/HelpGeneration/HelpGeneration.psm1 [hunk 8]
```diff
@@ -0,0 +1,343 @@
...
+    Param(
+        [Parameter(Mandatory = $true, Position = 0)]
+        [String]$HelpFolderPath
+    )
+
+    PROCESS
+    {
+        $HelpFolder = Get-Item $HelpFolderPath
+
+        $Exceptions = Import-Csv "$PSScriptRoot\..\..\src\Package\Exceptions\ValidateHelpExceptions.csv" | where { $_.Module -eq "$($HelpFolder.Parent.Name)" }
```
Comment (markcowl) [inline tools/HelpGeneration/HelpGeneration.psm1]: I think this should be a parameter

File: tools/HelpGeneration/HelpGeneration.psm1 [hunk 9]
```diff
@@ -0,0 +1,343 @@
...
+                        $errors += "$($HelpFolder.Parent.Name),$file,$error"
+                    }
+                }
+            }
+        }
+
+        # If there were any errors recorded, print them out and throw
+        if ($errors.Count -gt 0)
+        {
+            $errors | foreach { Add-Content "$PSScriptRoot\..\..\src\Package\ValidateHelpExceptions.csv" $_ }
```
Comment (markcowl) [inline tools/HelpGeneration/HelpGeneration.psm1]: I think this outupe file should be a parameter.  Also, I think you want this under ```src\Package\{buildconfig}``` by default in order to get it included in the build.  Finally, this should be called ValdateHelpIssues.csv to be consistent with the other products of static analysis.

File: tools/HelpGeneration/HelpGeneration.psm1 [hunk 10]
```diff
@@ -0,0 +1,343 @@
+﻿#Requires -Modules platyPS
```
Comment (markcowl) [inline tools/HelpGeneration/HelpGeneration.psm1]: Same comment about extra vertical space, this applies throughout

File: setup/azurecmdfiles.wxi (hunk 6 above)
Comment (cormacpayne) [inline setup/azurecmdfiles.wxi]: @markcowl all dll-Help.xml references are still in the wxi file - the ones being removed are from projects with dependencies on other projects (_e.g.,_ Profile, Tags, Resources)

**Comment by cormacpayne** [comment]
Sign job: http://azuresdkci.cloudapp.net/view/1-AzurePowerShell/job/powershell-sign/969/

**Comment by cormacpayne** [comment]
@azuresdkci test this please

**Pull request merged**